### PR TITLE
Refactor NodeCollection

### LINKED ISSUE #2203: Refactor NodeCollection
- From https://github.com/nest/nest-simulator/pull/2181#discussion_r741700472: `nc_const_iterator::operator*()` is quite complex and should be moved to node_collection.cpp. The same may be true for other functions as well.
- From https://github.com/nest/nest-simulator/pull/2181#discussion_r741701606: Can we always initialise `end_offset_` and `end_part_` in `nc_const_iterator`, to avoid having to check them against zero all the time?

### PR DESCRIPTION
**Pull request by hakonsbm**
Refactored the NodeCollection C++ classes to modernize, increase readability, and improve performance. Fixes #2203.

### CODE REVIEW

File: nestkernel/node_collection.cpp [hunk 1]
```diff
@@ -146,16 +206,16 @@ NodeCollection::NodeCollection()
 NodeCollectionPTR
 NodeCollection::create( const IntVectorDatum& node_idsdatum )
```
Comment (nicolossus) [inline nestkernel/node_collection.cpp:207]:
```suggestion
NodeCollection::create( const IntVectorDatum& node_ids_datum )
```

File: nestkernel/node_collection.cpp [hunk 2]
```diff
@@ -146,16 +206,16 @@ NodeCollection::NodeCollection()
...
 {
-  if ( node_idsdatum->size() == 0 )
+  if ( node_idsdatum->empty() )
```
Comment (nicolossus) [inline nestkernel/node_collection.cpp:209]:
```suggestion
  if ( node_ids_datum->empty() )
```

File: nestkernel/node_collection.cpp [hunk 3]
```diff
@@ -146,16 +206,16 @@ NodeCollection::NodeCollection()
...
   {
     return NodeCollection::create_();
   }
 
   std::vector< index > node_ids;
   node_ids.reserve( node_idsdatum->size() );
```
Comment (nicolossus) [inline nestkernel/node_collection.cpp:215]:
```suggestion
  node_ids.reserve( node_ids_datum->size() );
```

File: nestkernel/node_collection.cpp [hunk 4]
```diff
@@ -146,16 +206,16 @@ NodeCollection::NodeCollection()
...
-  for ( std::vector< long >::const_iterator it = node_idsdatum->begin(); it != node_idsdatum->end(); ++it )
+  for ( const auto& datum : *node_idsdatum )
```
Comment (nicolossus) [inline nestkernel/node_collection.cpp:216]:
```suggestion
  for ( const auto& datum : *node_ids_datum )
```

**Review by nicolossus — APPROVED** [review]
Thanks, overall this looks good to me! I have only have a few nitpicks you can consider.

**Review by jstapmanns — COMMENTED** [review]
Thank you very much for refactoring the NodeCollection code, @hakonsbm! I approve the code and while understanding it, I have come to appreciate the features of more modern c++ versions. My only comment concerns the tests for the NodeCollection. Are there scripts that test the slicing and the avoidance of adding overlapping NodeCollections? I think these would test nontrivial parts of the code.

**Comment by hakonsbm** [comment]
@jstapmanns That is tested in test_NodeCollection.py:
Test of slicing: https://github.com/nest/nest-simulator/blob/1a3e6f8ec30a066cedeea88a819d684199ed8ea6/testsuite/pytests/test_NodeCollection.py#L145

Test of adding overlapping NCs:
https://github.com/nest/nest-simulator/blob/1a3e6f8ec30a066cedeea88a819d684199ed8ea6/testsuite/pytests/test_NodeCollection.py#L304-L309

**Comment by jstapmanns** [comment]
Sorry, I see. I missed these. Then I approve the pull request.

**Review by jstapmanns — APPROVED** [review]

**Pull request merged**